### ISSUE TITLE
Sound Screen: Play a sine wave (when selected via checkbox) with the correct tone

### ISSUE THREAD

**Issue by samreid** [phetsims/wave-interference#373]
Request from the google doc:

> A checkbox under graph that says “Play sound” which does the tone and that will be the source sin frequency + amplitude, like Java.  But not having its own volume control.

**Comment by samreid** [comment]
I added a rough draft for a sine wave + checkbox in the preceding commits.  @jbphet can you please review and give feedback?

Aside from general review feedback, I had some questions in particular (also marked with TODOs):
- [ ] I noticed the sound sounds buggy if I put `{connectImmediately: true}`, can you help me understand what is happening in that case?
- [ ] I passed all `enableControlProperties` to initially be false, but when the checkbox is clicked, there is a short (<0.2sec) burst of audio.  I worked around it, but I wasn't sure why the workaround is necessary.

**Comment by samreid** [comment]
Also, there is a brief pitch change on reset all.  I thought this patch would eliminate it, but it didn't:

```
Index: js/waves/view/WavesScreenView.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/waves/view/WavesScreenView.js	(revision 321c88e93377e00f103a93a24b43b54af34f2ba8)
+++ js/waves/view/WavesScreenView.js	(date 1556057794000)
@@ -150,9 +150,14 @@
         x: this.waveAreaNode.left,
         centerY: this.waveAreaNode.centerY
       } );
+      const isNotResettingProperty = new BooleanProperty( true );
 
       const resetAllButton = new ResetAllButton( {
-        listener: () => model.reset(),
+        listener: () => {
+          isNotResettingProperty.value = false;
+          model.reset();
+          isNotResettingProperty.value = true;
+        },
         right: this.layoutBounds.right - MARGIN,
         bottom: this.layoutBounds.bottom - MARGIN
       } );
@@ -517,7 +522,8 @@
           enableControlProperties: [
             this.model.soundScene.isSoundPlayingProperty,
             this.model.soundScene.button1PressedProperty,
-            this.model.isRunningProperty
+            this.model.isRunningProperty,
+            isNotResettingProperty
           ]
         } );
         soundManager.addSoundGenerator( sineWavePlayer );
```

**Comment by jbphet** [comment]
Responses:
> I noticed the sound sounds buggy if I put {connectImmediately: true}, can you help me understand what is happening in that case?

The `connectImmediately` flag is intended to bypass the sound manager and connect directly to the audio context for easy testing of a sound generator during development, and isn't intended for use in production code.  The documentation for this parameter in `SoundGenerator` is intended to convey this, but if it didn't get the point across, I could use some help from you in how to make it more clear.  Bottom line: You shouldn't try to use it in this context.

> I passed all enableControlProperties to initially be false, but when the checkbox is clicked (even when the green speaker button is off) , there is a short (<0.2sec) burst of audio. I worked around it, but I wasn't sure why the workaround is necessary.

This is an ongoing issue and behaves differently in different browsers, see https://github.com/phetsims/tambo/issues/61.  I'd be interested to know which OS+Browser combos you saw this on (and it's fine if it is all of them).  I think implementing the suggestions below will likely solve this problem.

> Even when sound is disabled, there is a message that says: `phetAudioContext.js:129 The AudioContext was not allowed to start. It must be resumed (or created) after a user gesture on the page. https://goo.gl/7K7WLu` . This is from the module being loaded.  Is that as expected?

Yes, this is expected.  The message is Chrome's way of saying, "Hey, you're doing stuff with the audio context, but you're not going to hear anything until some user interaction occurs and you subsequently enable the audio context."  Chrome recently went through a number of changes to their autoplay policy, and I think the warning is intended to make sure developers are aware of how it works.  I expect that the warning will go away after some period of time, once everyone know that this is the autoplay policy from here on out.  The other browsers are following suit, though they all do it somewhat differently, but as of now tambo handles them all.

There are a couple of other pieces of feedback that I have about this sound generator.  The first one is that, as counter-intuitive as it seems, the Web Audio API wants you to start and stop oscillators rather than leaving them running and enabling/disabling them.  I suspect that this is because it consumes some system resources to leave them running.  In [this Stack Overflow entry](https://stackoverflow.com/questions/15261030/web-audio-start-and-stop-oscillator-then-start-it-again) the answerer describes this, and I've come across quite a number of other postings that essentially say the same thing.  Also, due to the auto-play policy, some of the browsers may not support attempts to start the oscillator at all during startup, which might explain some of the glitchiness that you were hearing (more on this below).

So, the general usage pattern that I had envisioned was one in which the sound generator is created and then explicitly turned on and off in something line a property change handler, not enabled/disabled through the enable control properties.  For an example, check out the `tambo` file `UIComponentsScreenView` and search for `sliderIncreaseClickSoundClip.play()`.  You're basically the first user of this library outside myself, so if you found that confusing, let's talk about how to make it clear.

Another unfortunate consequence of the autoplay policy is that if the very first action that the user takes would cause the sine wave sound to be generated, it may not start well, because the audio context may not have transitioned to the "running" state.  This might not be a problem in your particular use case, but if it is, we should talk.  For a preview of how to handle this, you can look at the code for staring a sound after the audio context transitions to the "running" state in the `tambo` file/method `NoiseGenerator.start()`.  It may make sense to pull this handling of delayed start into a common base class.

I'm up for collaborating on this if you'd like, but also feel free to continue logging questions in this issue and assigning it to me if that works better.

**Comment by samreid** [comment]
>  I'd be interested to know which OS+Browser combos you saw this on (and it's fine if it is all of them).

I was testing on Mac/Chrome.

**Comment by samreid** [comment]
> Another unfortunate consequence of the autoplay policy is that if the very first action that the user takes would cause the sine wave sound to be generated, it may not start well, because the audio context may not have transitioned to the "running" state. 

Would it be valuable to rewrite tambo so that no audio contexts are created until after the first user input?

**Comment by samreid** [comment]
> So, the general usage pattern that I had envisioned was one in which the sound generator is created and then explicitly turned on and off in something line a property change handler, not enabled/disabled through the enable control properties.

It appears `OscillatorNode` cannot be restarted.  Are you suggesting to create new `OscillatorNode` when the audio is restarted?

**Comment by jbphet** [comment]
> Would it be valuable to rewrite tambo so that no audio contexts are created until after the first user input?

I considered this but decided against for two main reasons.  The first is because I think it would be complex to defer **all** audio context manipulations until the audio context starts, and so far this hasn't been necessary, the only consequence being a pretty minor issue with playing a sound based on the very first action that the user takes (and this may go away as browser support for Web Audio becomes more consistent).   The second reason is that I think this would cause performance issues because a whole bunch of deferred audio initialization would need to be done based on the user's first interaction with the sim, which seems like a bad time to incur a delay.  Admittedly, I haven't ever tested this due to the challenges of doing so.

> It appears `OscillatorNode` cannot be restarted. Are you suggesting to create new `OscillatorNode` when the audio is restarted?

Yes, precisely.  I know that seems weird, and it has always struck me that way too, but apparently the Web Audio API is designed to make creating an oscillator a very fast operation.  I have done this in many tests and it works fine.  An `AudioBufferSource` is basically the equivalent of an oscillator, and you can look at `tambo` 'NoiseGenerator.start()` to see an example of a new one being created every time the sound is restarted.

**Comment by samreid** [comment]
@jbphet I updated it to recreate oscillators, can you please check that the implementation is correct, particularly about order of stop/disconnect, and connect/start or if there are other issues?

**Comment by samreid** [comment]
> The second reason is that I think this would cause performance issues because a whole bunch of deferred audio initialization would need to be done based on the user's first interaction with the sim, which seems like a bad time to incur a delay. Admittedly, I haven't ever tested this due to the challenges of doing so.

Doesn't that happen anyways because the audio context cannot start until the user input?  Or is there other overhead you are thinking of?

**Comment by jbphet** [comment]
> Doesn't that happen anyways because the audio context cannot start until the user input? Or is there other overhead you are thinking of?

From what I can glean from the behavior that I see and hear on the sound implementations done thus far, the browsers allow you to do things like creating sound sources and setting gain values with an audio context that is in the `suspended` state, it just doesn't allow any sound to actually be produced, and the audio context doesn't "run", meaning the `audioContext.currentTime` doesn't increase.  When the audio context makes the transition from `suspended` to `running` these things all start working, kind of like hooking up all the wires and then turning on the power.  I believe it would be more complicated to defer *all* initialization of the audio context - including its creation - until user interaction occurred, and my somewhat educated guess is that this would cause a noticeable delay between the first user interaction and the first sound, though as I stated before, I haven't actually tested this.  I also don't think that this is the intent of the autoplay policy, since it just came into existence recently and lots of people had written code that used Web Audio, and this sort of change would mean a pretty major rewrite for them as well as for us.

**Comment by jbphet** [comment]
I just reviewed `SineWaveGenerator.js`, and while the approach isn't what I've been doing, it looks good to me.  I've been using the enable controls just to set the overall enable state, and then using explicit methods to start and stop the sounds, so the approach of using additional enable properties to the constructor and linking to the `fullyEnabledProperty` and triggering the sound directly from there is a bit outside the box for me, but it was a box of my own making.  I like this approach, and will probably start to use it myself.  So, as long as it works, let's consider it reviewed and approved.

**Comment by samreid** [comment]
Thanks for the discussion and review @jbphet, I'm inclined to close this issue and continue in https://github.com/phetsims/wave-interference/issues/384 and https://github.com/phetsims/wave-interference/issues/386
